src/OutputWriters/netcdf_output_writer.jl

```diff
@@ -28,21 +28,21 @@ dictify(outputs::NamedTuple) = Dict(string(k) => dictify(v) for (k, v) in zip(ke
 
 Base.length(::Colon) = Inf
 
-xdim(::Type{Face}, inds) = length(inds) == 1 ? () : ("xF",)
-ydim(::Type{Face}, inds) = length(inds) == 1 ? () : ("yF",)
-zdim(::Type{Face}, inds) = length(inds) == 1 ? () : ("zF",)
+xdim(::Type{Face}, inds, user_inds) = (length(inds) == 1) & (length(user_inds) != 1) ? () : ("xF",)
+ydim(::Type{Face}, inds, user_inds) = (length(inds) == 1) & (length(user_inds) != 1) ? () : ("yF",)
+zdim(::Type{Face}, inds, user_inds) = (length(inds) == 1) & (length(user_inds) != 1) ? () : ("zF",)
```

Comment:
What are you trying to express here?

Reply:
If a sliced center field has indices `(:, 3,:)` then this doesn't define the field on dimesions `(xC, yC, zC)`, rather it defines it on dimensions `(xC, zC)`. Similarly for fields in other locations. This is a somewhat contrived way to make sliced fields work for NetCDF output, but only when the sliced dimension has length one (which I think is most cases). 

After reading a bit I think there's no good way to make arbitrarily sliced fields work for the NetCDF writer (which is mostly due to the limitations of the NetCDF format, not our code).

Reply:
okay, so there's no way to give it dimensions `(xC, yC[3], zC)`?

Reply:
Basically I think if it can work the ideal solution might extend `xnodes, ynodes, znodes` so that

```julia
x = xnodes(field)
```

is sufficient to get dimensions for NetCDF. I think right now we have

```julia
xnodes(f::Field) = xnodes(location(field), field.grid)
```

but we need `xnodes` to correctly slice the indices.

Reply:
But we can do the hacky thing here and the better solution via `xnodes` in another PR (if you think this `xnodes` solution will work).

Reply:
I think using `xnodes` is preferred just because it would probably make for cleaner code in general, but it wouldn't solve the problem posed by this PR.  To elaborate a bit more:

> okay, so there's no way to give it dimensions `(xC, yC[3], zC)`?

The only way to give a variable dimensions `(xC, yC[3], zC)` is if all the other variables have the same dimensions. Basically I can't define `var1` with dimensions `(xC, yC[1:128], zC)` and `var2` with dimensions `(xC, yC[3], zC)`. In this case `var2` _has_ to have dimensions `(xC, yC[1:128], zC)` too. One thing we can do is to only populated the `yC=3` with actual data and fill the rest with nans or something, but that's super wasteful in terms of storage. Which is why the solution here is to define `var2` on dimensions `(xC, zC)` only when that happens.

Another solution is to create another separate dimension `yC_sliced` that has only one value (`yC[3]`), but I'm not sure that's a good way to go.

Reply:
What if we create different dimensions for each `var`?

Reply:
Basically, rather than 3 dimensions, we need 3 + 3N dimensions, where `N` is the number of outputs that have "different" dimensions other than the ones associated with `grid` + `output_writer_indices`.

A side note, which probably should be done at the same time, is that we need to generalize beyond the name "xC" to the curvilinear `xᶜᶜᶜ`.

Reply:
> What if we create different dimensions for each `var`?

We could, but we need to be ready for a potentially high number of different NetCDF dimensions that refer back to the same dimension in Oceananigans (which might be confusing to users). I don't oppose that, but if we're gonna go that way, I think it's best to do re-write `NetCDFOutputWriter` to some extent (maybe on a different PR) since the way it's programmed right now is completely geared towards one set of dimensions to rule them all.


> A side note, which probably should be done at the same time, is that we need to generalize beyond the name "xC" to the curvilinear `xᶜᶜᶜ`, `xᶜᶠᶜ`, etc.

Yeah that's something I was planning on discussing and potentially changing here. The downside of the current Oceananigans nomenclature is that most non-Julia programmers wouldn't know how to type supercripts and subscripts on their editor...

Reply:
> We could, but we need to be ready for a potentially high number of different NetCDF dimensions that refer back to the same dimension in Oceananigans (which might be confusing to users)

I suggest using a design that creates a set of "defaults" that correspond to the `indices` passed into the `OutputWriter`. Then, when a field is encountered that has _different_ indices than the output writer indices, we create a new dimension. This avoids creating a large number of dimensions in most cases.

> The downside of the current Oceananigans nomenclature is that most non-Julia programmers wouldn't know how to type supercripts and subscripts on their editor...

We can use `x_ccc`, or add to https://github.com/CliMA/Oceananigans.jl/wiki/Installation-and-getting-started-with-Oceananigans#editing-and-reading-unicode-text-in-jl-files.